### PR TITLE
Electricity demand

### PR DESCRIPTION
**Pull request by jnnr**
demand

This PR takes euro-calliopes rules for creating electricity demand and casts them into a module.

### CODE REVIEW

File: modules/demand_electricity/LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) , AUTHORS
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (irm-codebase) [inline modules/demand_electricity/LICENSE:5]: Please update to 
```
Copyright (c) 2018-2023 Tim Tröndle (ETH Zürich / IASS Potsdam), Bryn Pickering (ETH Zürich)
Copyright (c) 2024 AUTHORS
```

File: modules/demand_electricity/README.md
Comment (irm-codebase) [inline modules/demand_electricity/README.md:1]: You might want to add the output directory info to this file

**Comment by jnnr** [comment]
> @jnnr seems like the current code is incompatible with v0.7 (demand cannot be negative). Would it be possible to alter the code slightly to produce positive outputs as part of this PR?

👍 Done in [1b666ba](https://github.com/calliope-project/ec_modules/pull/34/commits/1b666ba8aa4c19f9b573f300e1e79f070caa9610).

File: modules/demand_electricity/workflow/Snakefile [hunk 2]
```diff
@@ -18,7 +18,10 @@ with open(workflow.source_path("internal/internal_config.yaml"), "r") as f:
...
 
-if config["use_default_customisable_resources"]:
+if config["use_default_user_resources"]:
     rule all:
         message: "Generate all outputs for 'demand_electricity'."
-        input: "results/demand-electricity-ehighways.csv"
+        input:
+            expand("results/demand_electricity_{resolution}.csv", resolution="ehighways"),
+            expand("results/plot_timeseries_{resolution}.png", resolution="ehighways"),
+            expand("results/plot_map_{resolution}.png", resolution="ehighways")
```
Comment (irm-codebase) [inline modules/demand_electricity/workflow/Snakefile:27]: This is fine for now, but in the future we should move to a small check, so we can use this to generate unit tests.
https://snakemake.readthedocs.io/en/stable/snakefiles/testing.html

File: modules/demand_electricity/workflow/envs/plot.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# And cause a problem with matplotlib=3.7.1
+# Once geo.yaml is updated, this can env can probably be removed
+name: plot
+channels:
+    - conda-forge
+dependencies:
+  - python=3.11
+  - pandas
+  - geopandas
+  - matplotlib
```
Comment (irm-codebase) [inline modules/demand_electricity/workflow/envs/plot.yaml:11]: please freeze the version of the libraries!

**Review by irm-codebase — APPROVED** [review]
Approved!
A small improvement would be to freeze the version of the libraries used for plotting, so that they do not turn unstable in the future.

File: modules/demand_electricity/workflow/envs/plot.yaml (hunk 3 above)
Comment (jnnr) [reply modules/demand_electricity/workflow/envs/plot.yaml:11]: Do we want to freeze all environments managed by snakemake? Seems ok, just have to remove all restrictions once in a while, re-run, fix all bugs that may appear and freeze again.

**Pull request merged**